File: problems/valid_parenthesis.rb
# @param {String} s
# @return {Boolean}
def is_valid(s)
  stack = []
  s.each_char do |c|
    if c == '('
      stack.push(')')
    elsif c == '{'
      stack.push('}')
    elsif c == '['
      stack.push(']')
    elsif stack.empty? || c != stack.pop
      return false
    end
  end
  stack.empty?
end

# for char in s:
#   if char in dict.values():
#       stack.append(char)
#   elif char in dict.keys():
#       if stack == [] or dict[char] != stack.pop():
#           return False
#   else:
#       return False
# return stack == []

p is_valid("()")

File: strings/valid_parenthesis.py
class Solution:
    def isValid(self, s: str) -> bool:
        parens = {'{': '}', '(': ')', '[': ']'}
        stack = []

        for c in s:
            if c in parens: stack.append(parens[c])
            else:
                if not stack or stack.pop() != c: return False

        return stack == []


# 20. Valid Parentheses
# https://leetcode.com/problems/valid-parentheses/submissions/

# Push the corresponding closing brace onto stack each time we get an opening brace
# Each time we get a closing brace check match with top of stack and return false if not a match

# Time: O(n)
# Space: O(1)